src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/AccessModifierPlugin.kt

```diff
@@ -0,0 +1,44 @@
+package org.jenkinsci.gradle.plugins.accmod
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.attributes.Usage
+import org.gradle.api.plugins.JavaLibraryPlugin
+import org.gradle.api.tasks.compile.AbstractCompile
+import org.gradle.kotlin.dsl.apply
+import org.gradle.kotlin.dsl.named
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.withType
+
+open class AccessModifierPlugin : Plugin<Project> {
+    override fun apply(target: Project) {
+        with(target) {
+            apply<JavaLibraryPlugin>()
+            val library = dependencies.create("org.kohsuke:access-modifier-checker:1.21")
+            val jenkinsAccessModifier = configurations.create("jenkinsAccessModifier") {
+                attributes {
+                    attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.JAVA_RUNTIME))
+                }
+                isVisible = false
+                isCanBeConsumed = false
+                isCanBeResolved = true
+                withDependencies {
+                    add(library)
+                }
+            }
+
+            val propertyProvider = provider(PrefixedPropertiesProvider(this, CheckAccessModifierTask.PREFIX))
+            val checkAccessModifier = tasks.register<CheckAccessModifierTask>(CheckAccessModifierTask.NAME) {
+                dependsOn("classes")
+                val dirs = tasks.withType<AbstractCompile>().map { it.destinationDir }
```

Comment:
This will _realize_ all those tasks. Not necessarily a big deal but you may want to use a `provider {}` here to compute it lazily. This will also analyze test classes depending on the order Gradle decides to run the tasks because the `dirs` here doesn't have implicit task dependencies, and your explicit `dependsOn` only _selects_ part of the tasks.
If your task is run before `compileTestJava` for instance (or if that task is not run), it'll still be realized and your task will receive its `destinationDir`, which could be inexistant if the task never ran before (or could be stale).

You could use `SourceSet`s' `output.classesDirs`, which I believe will have implicit task dependencies attached to it, so `sourceSets["main"].output.classesDirs` will both implicitly declare task dependencies on `compileJava`, `compileGroovy`, and/or `compileKotlin` depending on the plugins applied to the project, **and** collect their destination directories; and without accidentally pulling in test tasks.

Reply:
Really appreciate this input! Great point about the test outputs. I've changed to it. This style also eliminates the need for checking if the directory exists.

I did have to do something a little different to get this to work:
```kotlin
project.extensions.getByType<SourceSetContainer>().getByName("main").output.classesDirs
```

I think this is the spirit of what you mentioned with this, but I believe I cannot resolve that `sourceSets` from the target project within a plugin:
```
sourceSets["main"].output.classesDirs
```

Reply:
I think you could also write something like

```kotlin
project.the<SourceSetContainer>()["main"].output.classes
```